codalab/worker_manager/slurm_batch_worker_manager.py

```diff
@@ -0,0 +1,268 @@
+import argparse
+import logging
+import uuid
+import subprocess
+import getpass
+from pathlib import Path
+import random
+import re
+
+from .worker_manager import WorkerManager, WorkerJob
+
+logger = logging.getLogger(__name__)
+
+
+class SlurmBatchWorkerManager(WorkerManager):
+    NAME = 'slurm-batch'
+    DESCRIPTION = 'Worker manager for submitting jobs using Slurm Batch'
+
+    SRUN_COMMAND = 'srun'
+    SBATCH_COMMAND = 'sbatch'
+    SBATCH_PREFIX = '#SBATCH'
+    SQUEUE_COMMAND = 'squeue'
+    SBATCH_COMMAND_RETRUN_REGEX = re.compile(r'^Submitted batch job (\d+)$')
+
+    @staticmethod
+    def add_arguments_to_subparser(subparser):
+        subparser.add_argument(
+            '--job-name',
+            type=str,
+            default='codalab-slurm-worker',
+            help='Name for the job definitions that will be generated by this worker manager',
+        )
+        subparser.add_argument(
+            '--nodelist', type=str, default='', help='The worker node to run jobs in'
+        )
+        subparser.add_argument(
+            '--partition', type=str, default='jag-standard', help='Name of batch job queue to use'
+        )
+        subparser.add_argument(
+            '--cpus', type=int, default=1, help='Default number of CPUs for each worker'
+        )
+        subparser.add_argument(
+            '--gpus', type=int, default=1, help='Default number of GPUs for each worker'
+        )
+        subparser.add_argument(
+            '--memory-mb', type=int, default=2048, help='Default memory (in MB) for each worker'
+        )
+        subparser.add_argument(
+            '--dry-run',
+            action='store_true',
+            help='Print out Slurm batch job definition without submitting to Slurm',
+        )
+        subparser.add_argument(
+            '--user', type=str, default=getpass.getuser(), help='User to run the Batch jobs as'
+        )
+        subparser.add_argument(
+            '--password-file',
```

Comment:
can you make this optional and also check for the `CODALAB_USERNAME` and `CODALAB_PASSWORD` environment variables? Like how the AWSBatchWorkerManager does it: https://github.com/codalab/codalab-worksheets/blob/master/codalab/worker_manager/aws_batch_worker_manager.py#L108-L109

Reply:
Ah, I was looking for this!!! That would be great to have this from ENV. I am updating.

Reply:
Also, this would require some manual setup to add both environment variables in your environment.

Reply:
that's fine. I think you should make `--password-file` optional. If password-file is provided and the environment variables are set, default to the password-file. If password-file is not provided and the environment variables are set, use them. If the environment variables aren't set, and password-file isn't specified, raise an error?

Reply:
Yup, I think that works! I think I need to check the existence of `CODALAB_USERNAME` and `CODALAB_PASSWORD` early so that the newly spawned worker won't keep hanging there or failed inside the worker without giving any notice to the user.